### PR TITLE
VMware sdk poc

### PR DESCRIPTION
**Pull request by mihaitusa**
Virtual machines - createVM, deleteVM, listVMsAndTemplates, getVMDetails, powerOnVM, powerOffVM, updateVM; 
Virtual machines/conf - GetOsDescriptors as helping operation needed for createVM.

### CODE REVIEW

File: score-vmware/src/main/java/io/cloudslang/content/actions/vm/CreateVM.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Created by Mihai Tusa.
+ * 10/18/2015.
+ */
+public class CreateVM {
+
+    private VmService service = new VmService();
```
Comment (gituser173) [inline score-vmware/src/main/java/io/cloudslang/content/actions/vm/CreateVM.java]: maybe use the interface and not the implementation of the service here

File: score-vmware/src/main/java/io/cloudslang/content/actions/vm/CreateVM.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+        int intPort = InputUtils.getIntInput(port, Inputs.DEFAULT_HTTPS_PORT);
+        boolean booleanX509HostnameVerifier = InputUtils.getBooleanInput(x509HostnameVerifier);
+        int intNumCPUs = InputUtils.getIntInput(numCPUs, Constants.DEFAULT_CPU_COUNT);
+        long longVmDiskSize = InputUtils.getLongInput(vmDiskSize, Constants.DEFAULT_VM_DISK_SIZE_MB);
+        long longVmMemorySize = InputUtils.getLongInput(vmMemorySize, Constants.DEFAULT_VM_MEMORY_SIZE_MB);
+
+        try {
+
+            ManagedObjectReference managedObjectReference = service.createVirtualMachine(host,
```
Comment (gituser173) [inline score-vmware/src/main/java/io/cloudslang/content/actions/vm/CreateVM.java]: the service could return the resultMap, we should not know about the existence of ManagedObjectReference at this point
also the createVirtualMachine method has too many parameters

File: score-vmware/src/main/java/io/cloudslang/content/vmware/services/VmService.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Created by Mihai Tusa.
+ * 1/6/2016.
+ */
+public class VmService {
```
Comment (gituser173) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/services/VmService.java]: maybe break this Object as it starts to become large and is doing too much

File: score-vmware/src/main/java/io/cloudslang/content/vmware/actions/vm/CreateVM.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                    virtualMachineName,
+                    description,
+                    dataStore,
+                    numCPUs,
+                    vmDiskSize,
+                    vmMemorySize,
+                    guestOsId);
+
+            VmService vmService = new VmService();
+            resultMap = vmService.createVirtualMachine(httpInputs, vmInputs);
```
Comment (gituser173) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/actions/vm/CreateVM.java]: inline where you use the variable only once

File: score-vmware/src/main/java/io/cloudslang/content/vmware/constants/ErrorMessages.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package io.cloudslang.content.vmware.constants;
+
+/**
+ * Created by Mihai Tusa.
+ * 10/20/2015.
+ */
+public class ErrorMessages {
```
Comment (gituser173) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/constants/ErrorMessages.java]: maybe use enum instead for the constants as these classes tend to get big in size over time

File: score-vmware/src/main/java/io/cloudslang/content/vmware/entities/http/Protocol.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    HTTPS;
+
+    public static String getValue(String input) throws Exception {
+        if (StringUtils.isBlank(input)) {
+            return HTTPS.toString();
+        }
+        try {
+            return valueOf(input.toUpperCase()).toString();
+        } catch (IllegalArgumentException iae) {
+            throw new Exception(ErrorMessages.UNSUPPORTED_PROTOCOL);
```
Comment (Bonczidai) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/entities/http/Protocol.java]: you should include the name of the unsupported protocol in the error message

File: score-vmware/src/main/java/io/cloudslang/content/vmware/entities/http/Protocol.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * Created by Mihai Tusa.
+ * 10/30/2015.
+ */
+public enum Protocol {
+    HTTP,
+    HTTPS;
+
+    public static String getValue(String input) throws Exception {
+        if (StringUtils.isBlank(input)) {
```
Comment (Bonczidai) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/entities/http/Protocol.java]: is there a specific reason for throwing checked exceptions? if no, better use RuntimeException

File: score-vmware/src/main/java/io/cloudslang/content/vmware/utils/FindObjects.java [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+ * the different versions of findObject.  This was done (or not-done) deliberately
+ * in order to make it easier to follow the code, for learning purposes.
+ */
+
+/**
+ * Utility methods for working with Java Objects in VMware's vCenter Java libraries. These methods
+ * work mostly with Managed Object Reference objects and Object Content objects.
+ */
+
+public class FindObjects {
```
Comment (Bonczidai) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/utils/FindObjects.java]: if the utility class contains only static methods, it is a good idea to declare it static class or final + private constructor

File: score-vmware/src/main/java/io/cloudslang/content/vmware/connection/ConnectionResources.java [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+        setHostMor(vmInputs);
+        setComputeResourceMor();
+        setResourcePoolMor();
+        setVmFolderMor();
+
+        this.setVimPortType(connection.getVimPort());
+        this.setGetMOREF(getMOREF);
+        this.setConnection(connection);
+        this.setMorRootFolder(morRootFolder);
+        this.setServiceInstance(serviceInstance);
```
Comment (gituser173) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/connection/ConnectionResources.java]: these last few setters seem to be redundant

File: score-vmware/src/main/java/io/cloudslang/content/vmware/connection/ConnectionResources.java [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+    private void setDataCenterMor(VmInputs vmInputs) throws InvalidPropertyFaultMsg, RuntimeFaultFaultMsg {
+        ManagedObjectReference dataCenterMor = null;
+        if(StringUtils.isNotBlank(vmInputs.getDataCenterName())){
+            dataCenterMor = getdataCenterMor(vmInputs.getDataCenterName(), morRootFolder, getMOREF);
+        }
+        this.setDataCenterMor(dataCenterMor);
+    }
+
+    private ManagedObjectReference getdataCenterMor(String dataCenterName, ManagedObjectReference mor, GetMOREF getMOREF)
```
Comment (gituser173) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/connection/ConnectionResources.java]: typo

File: score-vmware/src/main/java/io/cloudslang/content/vmware/utils/FindObjects.java (hunk 8 above)
Comment (mihaitusa) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/utils/FindObjects.java]: @Bonczidai  I made the constructor private... I'll prefer to let the class at it is... maybe in the future will have some wrappers...

File: score-vmware/src/main/java/io/cloudslang/content/vmware/entities/http/Protocol.java (hunk 7 above)
Comment (mihaitusa) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/entities/http/Protocol.java]: @Bonczidai  You're right about it! I corrected it.

File: score-vmware/src/main/java/io/cloudslang/content/vmware/connection/ConnectionResources.java (hunk 10 above)
Comment (mihaitusa) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/connection/ConnectionResources.java]: @gituser173  You're right about setters. I corrected it.

File: score-vmware/src/main/java/io/cloudslang/content/vmware/connection/helpers/GetMOREF.java [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+                                                        .skip(Boolean.FALSE)
+                                                        .type(Constants.CONTAINER_VIEW)
+                                        )
+                        )
+        };
+    }
+
+    public RetrieveResult containerViewByType(final ManagedObjectReference container,
+                                              final String morefType,
+                                              final String[] morefProperties,
```
Comment (gituser173) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/connection/helpers/GetMOREF.java]: container, morefType, moreFProperties are not used here?

File: score-vmware/src/main/java/io/cloudslang/content/vmware/constants/ErrorMessages.java (hunk 5 above)
Comment (mihaitusa) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/constants/ErrorMessages.java]: @gituser173  It's worth waiting a little bit to see how much we will develop in VMware integration. If the class will grow bigger fast then I will break it in several Enum's  (e.g.: one enum for each group of errors).

File: score-vmware/src/main/java/io/cloudslang/content/vmware/connection/exceptions/ConnectionMalformedUrlException.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package io.cloudslang.content.vmware.connection.exceptions;
+
+import java.net.MalformedURLException;
+
+/**
+ * Created by Mihai Tusa.
+ * 10/20/2015.
+ */
+public class ConnectionMalformedUrlException extends ConnectionException {
```
Comment (gituser173) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/connection/exceptions/ConnectionMalformedUrlException.java]: looks like this isn't used

File: score-vmware/src/main/java/io/cloudslang/content/vmware/utils/InputUtils.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+    public static String getUrlString(HttpInputs httpInputs) throws Exception {
+        String protocolString = Protocol.getValue(httpInputs.getProtocol());
+        String urlString = protocolString + "://" + httpInputs.getHost() + ":" + httpInputs.getPort() + Constants.URI_PATH;
+        URL url = new URL(urlString.toLowerCase());
+
+        return url.toString();
+    }
+
+    public static String getResultsMap(Map<String, String> inputMap, String delimiter) {
```
Comment (gituser173) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/utils/InputUtils.java]: maybe use a JSON library to do a pretty print

File: score-vmware/src/main/java/io/cloudslang/content/vmware/services/VmService.java (hunk 3 above)
Comment (mihaitusa) [inline score-vmware/src/main/java/io/cloudslang/content/vmware/services/VmService.java]: @gituser173  I cleaned up more some methods... did the helpers/VmUtils

**Comment by Bonczidai** [comment]
score-vmware is ignored on Travis - updated the config file.. go ahead @mihaitusa :+1:

**Pull request merged**